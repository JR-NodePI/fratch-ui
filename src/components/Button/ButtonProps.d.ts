import { IconPlus as Icon } from '../Icon/Icons';
import { ButtonSize, ButtonType } from './ButtonConstants';

export interface ButtonProps {
  children?: string;
  className?: string;
  disabled?: boolean;
  Icon?: typeof Icon;
  isRound?: boolean;
  label?: string;
  onClick?: () => void;
  size?: (typeof ButtonSize)[keyof typeof ButtonSize];
  stretch?: boolean;
  type?: (typeof ButtonType)[keyof typeof ButtonType];
}
