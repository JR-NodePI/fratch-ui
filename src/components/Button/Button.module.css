.button {
  gap: 16px;
  display: inline-flex;
  align-items: center;
  border-radius: var(--ft-button-default-border-radius);
  border: none;
  letter-spacing: var(--ft-button-default-letter-spacing);
  font-family: var(--ft-button-default-font-family);
  font-size: var(--ft-button-default-font-size);
  font-weight: var(--ft-button-default-font-weight);
  max-width: 100%;
  overflow: hidden;
  padding: 0 var(--ft-button-default-padding-x);
  text-overflow: ellipsis;
  transition: box-shadow 0.2s ease-out, background-color 0.2s ease-out;
  white-space: nowrap;
  height: var(--ft-button-default-height);
  text-align: center;
}
.button.stretch {
  width: 100%;
}
.button.only_icon {
  border-radius: 100%;
  padding: 0;
  width: var(--ft-button-default-height);
  justify-content: center;
}
.button * {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  width: 100%;
}
.button:hover {
  box-shadow: var(--ft-box-shadow-2), inset 0 -1px 0 0 rgba(0, 0, 0, 0.15),
    inset 0 1px 0 0 rgba(255, 255, 255, 1);
  background-color: var(--ft-button-hover-bg-color);
  --icon-color: var(--ft-button-hover-color);
  color: var(--ft-button-hover-color);
}
.button:active {
  background-color: var(--ft-button-active-bg-color);
}
.button,
.button:disabled {
  background-color: var(--ft-button-default-bg-color);
  --icon-color: var(--ft-button-default-color);
  box-shadow: var(--ft-box-shadow-2), inset 0 -1px 0 0 rgba(0, 0, 0, 0.1),
    inset 0 1px 0 0 rgba(255, 255, 255, 0.7);
  color: var(--ft-button-default-color);
}
.button:disabled {
  filter: saturate(0.5) brightness(135%) contrast(80%);
  user-select: none;
}

.primary {
  background-color: var(--ft-primary-default-bg-color);
  color: var(--ft-primary-default-color);
  --icon-color: var(--ft-primary-default-color);
  box-shadow: 0 1px 4px 0 var(--ft-primary-shadow-color);
}
.primary:hover {
  background-color: var(--ft-primary-hover-bg-color);
  color: var(--ft-primary-hover-color);
  --icon-color: var(--ft-primary-hover-color);
  box-shadow: 0 3px 16px 0 var(--ft-primary-hover-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.229),
    inset 0 1px 0 0 rgba(0, 251, 255, 0.504);
}
.primary:active {
  background-color: var(--ft-primary-active-bg-color);
  color: var(--ft-primary-default-color);
  --icon-color: var(--ft-primary-default-color);
}
.primary,
.primary:disabled {
  background-color: var(--ft-primary-default-bg-color);
  color: var(--ft-primary-default-color);
  --icon-color: var(--ft-primary-default-color);
  box-shadow: 0 1px 4px 0 var(--ft-primary-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.3), inset 0 1px 0 0 rgba(0, 251, 255, 0.4);
}

.secondary:hover {
  background-color: var(--ft-secondary-hover-bg-color);
  color: var(--ft-secondary-hover-color);
  --icon-color: var(--ft-secondary-hover-color);
  box-shadow: 0 3px 16px 0 var(--ft-secondary-hover-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.2),
    inset 0 1px 0 0 rgba(255, 146, 244, 0.929);
}
.secondary:active {
  background-color: var(--ft-secondary-active-bg-color);
  color: var(--ft-secondary-default-color);
  --icon-color: var(--ft-secondary-default-color);
}
.secondary,
.secondary:disabled {
  background-color: var(--ft-secondary-default-bg-color);
  color: var(--ft-secondary-default-color);
  --icon-color: var(--ft-secondary-default-color);
  box-shadow: 0 1px 4px 0 var(--ft-secondary-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.3), inset 0 1px 0 0 rgb(255, 146, 244, 0.8);
}

.tertiary:hover {
  background-color: var(--ft-tertiary-hover-bg-color);
  color: var(--ft-tertiary-hover-color);
  --icon-color: var(--ft-tertiary-hover-color);
  box-shadow: 0 3px 16px 0 var(--ft-tertiary-hover-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.339),
    inset 0 1px 0 0 rgba(141, 179, 255, 0.815);
}
.tertiary:active {
  background-color: var(--ft-tertiary-active-bg-color);
  color: var(--ft-tertiary-default-color);
  --icon-color: var(--ft-tertiary-default-color);
}
.tertiary,
.tertiary:disabled {
  background-color: var(--ft-tertiary-default-bg-color);
  color: var(--ft-tertiary-default-color);
  --icon-color: var(--ft-tertiary-default-color);
  box-shadow: 0 1px 4px 0 var(--ft-tertiary-shadow-color),
    inset 0 -1px 0 0 rgba(0, 0, 0, 0.3),
    inset 0 1px 0 0 rgba(124, 224, 255, 0.6);
}
/***** size *****/
.small {
  gap: 12px;
  font-size: var(--ft-button-small-font-size);
  padding: 0 var(--ft-button-small-padding-x);
  height: var(--ft-button-small-height);
}
.small.only_icon {
  width: var(--ft-button-small-height);
}

.medium {
  gap: 16px;
  font-size: var(--ft-button-medium-font-size);
  padding: 0 var(--ft-button-medium-padding-x);
  height: var(--ft-button-medium-height);
}
.medium.only_icon {
  width: var(--ft-button-medium-height);
}

.large {
  gap: 20px;
  font-size: var(--ft-button-large-font-size);
  padding: 0 var(--ft-button-large-padding-x);
  height: var(--ft-button-large-height);
}
.large.only_icon {
  width: var(--ft-button-large-height);
}

/***** icon *****/
.button .icon {
  margin: -0.1em;
  font-size: 1.1em;
  width: 1.1em;
}
.button .icon [fill] {
  fill: var(--icon-color);
}
.button .icon [stroke] {
  stroke: var(--icon-color);
}
.button .icon [fill='none'] {
  fill: transparent;
}
.button .icon [stroke='none'] {
  stroke: transparent;
}
