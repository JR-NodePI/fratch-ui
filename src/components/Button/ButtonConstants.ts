export const ButtonType = {
  DEFAULT: 'default',
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
} as const;

export const ButtonSize = {
  LARGE: 'large',
  MEDIUM: 'medium',
  SMALL: 'small',
  SMALLER: 'smaller',
} as const;
