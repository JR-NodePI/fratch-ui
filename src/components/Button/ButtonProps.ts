export const ButtonType = Object.freeze({
  DEFAULT: 'default',
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  TERTIARY: 'tertiary',
});

export const ButtonSize = Object.freeze({
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large',
});

export interface ButtonProps {
  className?: string;
  label: string;
  onClick?: () => void;
  size?: (typeof ButtonSize)[keyof typeof ButtonSize];
  type?: (typeof ButtonType)[keyof typeof ButtonType];
}
