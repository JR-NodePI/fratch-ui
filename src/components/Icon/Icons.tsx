export { default as IconArrowDown } from './IconArrowDown';
export { default as IconClose } from './IconClose';
export { default as IconPlus } from './IconPlus';
export { default as IconCheck } from './IconCheck';
