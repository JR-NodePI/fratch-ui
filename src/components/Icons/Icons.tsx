export { default as IconArrowDown } from './Icons/IconArrowDown';
export { default as IconCheck } from './Icons/IconCheck';
export { default as IconClose } from './Icons/IconClose';
export { default as IconDownload } from './Icons/IconDownload';
export { default as IconEdit } from './Icons/IconEdit';
export { default as IconError } from './Icons/IconError';
export { default as IconExport } from './Icons/IconExport';
export { default as IconImport } from './Icons/IconImport';
export { default as IconInfo } from './Icons/IconInfo';
export { default as IconPause } from './Icons/IconPause';
export { default as IconPin } from './Icons/IconPin';
export { default as IconPlay } from './Icons/IconPlay';
export { default as IconPlus } from './Icons/IconPlus';
export { default as IconSuccess } from './Icons/IconSuccess';
export { default as IconUpload } from './Icons/IconUpload';
export { default as IconVerticalDots } from './Icons/IconVerticalDots';
export { default as IconWarning } from './Icons/IconWarning';
