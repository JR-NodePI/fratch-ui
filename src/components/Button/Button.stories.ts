import { ButtonSizes, ButtonTypes } from "./ButtonProps";
import Button from "./Button";
import type { Meta, StoryObj } from "@storybook/react";

const meta = {
  title: "Example/Button",
  component: Button,
  tags: ["autodocs"],
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    label: "DEFAULT",
    type: ButtonTypes.PRIMARY,
    size: ButtonSizes.MEDIUM,
  },
};
