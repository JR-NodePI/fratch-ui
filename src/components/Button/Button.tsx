import { c } from '../../helpers/classNameHelpers';

import { ButtonSize, ButtonType } from './ButtonProps';
import type { ButtonProps } from './ButtonProps';

import styles from './Button.module.css';

const Button = ({
  className,
  label,
  onClick,
  size = ButtonSize.MEDIUM,
  type = ButtonType.PRIMARY,
}: ButtonProps): JSX.Element => {
  return (
    <button
      className={c(styles.default, styles[type], styles[size], className)}
      onClick={onClick}
    >
      {label}
    </button>
  );
};

export default Button;
