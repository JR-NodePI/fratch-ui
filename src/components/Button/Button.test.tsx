import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import { IconCheck } from '../Icons/Icons';
import Button from './Button';
import { ButtonSize, ButtonType } from './ButtonConstants';
import { type ButtonProps } from './ButtonProps';

describe('Button', () => {
  const setup = (props: ButtonProps = {}): ReturnType<typeof render> =>
    render(<Button label="Test button" {...props} />);

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  describe('should render with SIZE', () => {
    Object.values(ButtonSize).forEach(size => {
      it(size, () => {
        const { container } = setup({ size });
        expect(container).toMatchSnapshot();
      });
    });
  });

  describe('should render with TYPE', () => {
    Object.values(ButtonType).forEach(type => {
      it(type, () => {
        const { container } = setup({ type });
        expect(container).toMatchSnapshot();
      });
    });
  });

  it('should render with children over label', () => {
    const { container } = setup({
      children: 'Test children',
      label: 'Test label',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with stretch', () => {
    const { container } = setup({
      stretch: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with isRound', () => {
    const { container } = setup({
      isRound: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with Icon', () => {
    const { container } = setup({
      Icon: IconCheck,
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with className', () => {
    const { container } = setup({
      className: 'mock_test_class_name',
    });
    expect(container).toMatchSnapshot();
  });

  it('should render with disabled', () => {
    const { container } = setup({
      disabled: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should perform onClick', async () => {
    const mockOnClick = vi.fn();

    setup({ label: 'mock-test-button', onClick: mockOnClick });

    const buttonElement = screen.getByRole('button', {
      name: /mock-test-button/i,
    });
    await userEvent.click(buttonElement);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
    expect(mockOnClick).toHaveBeenCalledWith(
      expect.objectContaining({
        type: 'click',
        target: buttonElement,
      })
    );
  });
});
