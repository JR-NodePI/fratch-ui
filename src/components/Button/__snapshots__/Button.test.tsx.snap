// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button.tsx > should render properly 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with Icon 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
  >
    <svg
      class="_icon_2be8fc _primary_2be8fc _icon_2a1efe"
      fill="none"
      viewBox="-1 -3 18 18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1 5.917 5.724 10.5 15 1.5"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="3"
      />
    </svg>
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with SIZE > large 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _large_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with SIZE > medium 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with SIZE > small 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _small_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with TYPE > default 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with TYPE > primary 1`] = `
<div>
  <button
    class="_button_2a1efe _primary_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with TYPE > secondary 1`] = `
<div>
  <button
    class="_button_2a1efe _secondary_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with TYPE > tertiary 1`] = `
<div>
  <button
    class="_button_2a1efe _tertiary_2a1efe _medium_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with children over label 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
  >
    Test children
  </button>
</div>
`;

exports[`Button.tsx > should render with className 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe mock_test_class_name"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with disabled 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe"
    disabled=""
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;

exports[`Button.tsx > should render with isRound 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe _only_icon_2a1efe"
    title="Test button"
  />
</div>
`;

exports[`Button.tsx > should render with stretch 1`] = `
<div>
  <button
    class="_button_2a1efe _default_2a1efe _medium_2a1efe _stretch_2a1efe"
  >
    <span>
      Test button
    </span>
  </button>
</div>
`;
